src/main/java/com/google/sps/servlets/ExamServlet.java

```diff
@@ -0,0 +1,171 @@
+
+  /**
+   * Gets the exam ID from the httpRequest, displays exam if it exists otherwise displays
+   * list of exams.
+   *
+   * @param request  provides request information from the HTTP servlet
+   * @param response response object where servlet will write information to
+   */
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Only logged in users should access this page.
+    UserService userService = UserServiceFactory.getUserService();
+    if (!userService.isUserLoggedIn()) {
+      logger.atWarning().log("User is not logged in.");
+      response.sendRedirect("/");
+      return;
+    }
+    logger.atInfo().log("user=%s", userService.getCurrentUser());
+    response.setContentType("text/html;");
+    PrintWriter out = response.getWriter();
+    out.println("<!DOCTYPE html>");
+    out.println("<html>");
+    out.println("<head>");
+    out.println("<link href=\"https://fonts.googleapis.com/css2?family=Domine:wght@400;"
+        + "700&family=Open+Sans:ital,wght@0,400;0,600;0,700;1,400;1,600;1,700&display=swap\""
+        + " rel=\"stylesheet\">");
+    out.println("<link rel=\"stylesheet\" href=\"style.css\">");
+    out.println("<script src=\"script.js\"></script>");
+    out.println("<script src=\"examSubmission.js\"></script>");
+    out.println("<title>Take Exam</title>");
+    out.println("</head>");
+    out.println("<body>");
+    out.println("<header>");
+    out.println("<div class=\"navtop\">");
+    out.println("<p><a href=\"index.html\">Homepage</a></p>");
+    out.println("<p><a href=\"dashboard.html\">Dashboard</a></p>");
+    out.println("<p id=logInOut></p>");
+    out.println("</div>");
+    out.println("</header>");
+    out.println("<main>");
+    String examID = UtilityClass.getParameter(request, "examID", null);
+    Entity examEntity = null;
+
+    if (examID != null) {
+      // If an exam has been selected
+      try {
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        Key key = KeyFactory.createKey("Exam", Long.parseLong(examID));
+        examEntity = datastore.get(key);
+      } catch (Exception e) {
+        out.println("<h3>Selected exam is not available.</h3>");
+        logger.atInfo().log("Exam ID does not exist: %s", e);
+      }
+      if (examEntity != null) {
+        // If exam exists, then display the exam and questions
+        String name = (String) examEntity.getProperty("name");
+        String duration = (String) examEntity.getProperty("duration");
+        final String ownerID = (String) examEntity.getProperty("ownerID");
+        List<Long> questionsList = null;
+        try {
+          questionsList = (List<Long>) examEntity.getProperty("questionsList");
+        } catch (Exception e) {
+          logger.atWarning().log("There was an error getting the questions list: %s", e);
+        }
+        out.println("<h1>Exam Name: " + name + "</h1>");
```

Comment:
I think this is an OK approach for the MVP, but overall try not to mix frontend representation and backend data extraction.

A common technique for that is representing your pages as templates and using backend logic to fill in concrete data into them. E.g. in Python it's common to use https://en.wikipedia.org/wiki/Jinja_(template_engine). I'm sure Java has similar libraries.

Reply:
Issue created referencing this.